### PR TITLE
[CUSFEES-12] Fix: Per-rule valuation and compound bases for CA/AU/NZ presets

### PR DESCRIPTION
**Pull request by Abdalsalaam**
### Description

The shipped Canada, Australia, and New Zealand customs presets were calculating import tax on the wrong base. These jurisdictions require **different valuation bases for duty vs. import tax within the same country** (e.g. Canada: duty on FOB, GST on CIF + Duty), but the plugin only exposed a single *global* FOB/CIF setting — so no configuration produced a compliant result. Merchants using these presets in good faith were under- or over-collecting tax at checkout.

This PR introduces **per-rule valuation** and **compound bases** so a single country preset can mix valuation methods and stack rules on top of other rules' outputs (e.g. tax base = CIF + Duty). The calculator is re-architected as a round-based engine with cycle detection so dependent rules resolve in the correct order. Existing presets (CA/AU/NZ/UK/EU) are updated to declare the correct per-rule valuation and dependencies, and a one-time migration assigns stable `rule_id`s and sane defaults to existing user rules.

### Changes in this PR

- Calculator rewritten as a round-based engine that resolves rule dependencies via `base_includes` and detects cycles before evaluating (`includes/class-cfwc-calculator.php`).
- Per-rule `valuation` (`fob` / `cif` / inherit-global) and `base_includes` (list of dependent `rule_id`s) added to the rule schema, with sanitization, migration, and CSV export/import support.
- Stable `rule_id`s introduced and preserved through the rule matcher, AJAX save flow, CSV import, and preset append (preset IDs are suffixed on append to prevent collisions).
- Admin UI: new **Valuation** and **Depends on** columns in the rules table, populated select options in `admin.js`, dependency error notice, and i18n strings.
- Presets updated (CA GST, AU GST, NZ GST, UK VAT, EU VAT) to use correct per-rule valuation and `base_includes` so import tax compounds on duty per CBSA / ABF / HMRC / EU rules.
- One-time migration hook for v1.2.0 gated on a dedicated option (not plugin version) and triggered on `init` so the frontend cart sees migrated data; rules transient and calculator cache are explicitly cleared on save.
- CSV import tolerates column-count drift from older exports.
- Documentation updated: `docs/CIF.md` covers per-rule valuation and compound bases; `changelog.txt` / `readme.txt` updated.

### Test Instructions

1. Checkout this branch locally and run `composer install && pnpm install && pnpm build:dev`.
2. Activate the plugin on a fresh WC store and go to **WooCommerce → Settings → Tax → Customs & Import Fees**.
3. Leave **Customs Valuation Method** on its default (FOB). Import the **Canada GST & Duty** preset.
4. Create a simple product priced at **$100** with any origin.
5. Add to cart, set the shipping destination to **Canada**, and add a shipping method totaling **$20**.
6. Go to checkout and inspect the customs fee breakdown. Expected:
   - **Canada Duty (Import):** `8% × $100 = $8.00` (FOB base)
   - **Canada GST (Import):** `5% × ($100 + $20 + $8) = $6.40` (CIF + Duty base)
7. Repeat with the **Australia GST** and **New Zealand GST** presets — import tax should compound on duty + freight, matching ABF/NZ rules.
8. Repeat with **UK VAT** and **EU VAT** — both duty and VAT use CIF; VAT should compound on duty.
9. Open **Rules** table and confirm the new **Valuation** and **Depends on** columns render and are editable. Try creating a rule that depends on itself — the dependency error notice should appear.
10. Export rules to CSV, re-import, and verify `valuation` / `base_includes` / `rule_id` are preserved. Try importing an older CSV (fewer columns) — it should still import without fatal.
11. On a site with pre-existing rules from v1.1.x, deploy this branch and load any wp-admin page; confirm the migration runs once (each rule gains a stable `rule_id` and default valuation), and that the frontend cart calculates fees correctly without a second page load.
12. Run **WooCommerce → Status → Logs** and confirm no new warnings/notices in `cfwc-*` logs during the flows above.

### Things to check

- [x] Are all texts internationalized?
- [x] Has a cache been added?
- [x] Are the hooks/filters called only when necessary?
- [x] Have the local logs been checked to ensure this PR does not introduce new errors, warnings, or notifications?

### CODE REVIEW

File: includes/admin/class-cfwc-ajax.php [hunk 1]
```diff
@@ -280,26 +338,72 @@ public function import_rules() {
 		$headers   = str_getcsv( array_shift( $lines ), ',', '"', '\\' );
 		$new_rules = array();
 
+		// Build header index map for flexible column mapping.
+		$header_map = array();
+		foreach ( $headers as $index => $header ) {
+			$header_map[ strtolower( trim( $header ) ) ] = $index;
+		}
+
```
Comment (dustinparker) [inline includes/admin/class-cfwc-ajax.php:401]: @Abdalsalaam, shouldn't we add an array of required headers that need to be populated in order to import a valid rule?

The settings save path at `class-cfwc-settings.php:265` already drops rules that have no `country` or `to_country`, but the importer at `class-cfwc-ajax.php:383` doesn't enforce the same thing. So it's possible to import a CSV with empty cells and end up with rules that just sit there inert.

File: includes/class-cfwc-calculator.php [hunk 2]
```diff
@@ -299,51 +297,148 @@ public function calculate_fees( $cart ) {
...
-							$fee
-						)
-					);
+				$pending = $still_pending;
+
+				if ( ! $made_progress ) {
+					// Unresolvable: cycle or external dependency.
+					// Compute remaining rules on their own base so the cart still works.
+					foreach ( $pending as $rule ) {
+						$rule_id = isset( $rule['rule_id'] ) && '' !== $rule['rule_id'] ? $rule['rule_id'] : 'rule_' . wp_rand( 1000, 9999 );
```
Comment (dustinparker) [inline includes/class-cfwc-calculator.php]: @Abdalsalaam should we use `wp_generate_uuid4()` here instead of `wp_rand( 1000, 9999 )`?

The 4-digit random space is small enough that two rules hitting this fallback in the same cart calc could collide, and the second one would silently overwrite the first in the `$calculated` map.

The rest of the codebase already uses `wp_generate_uuid4()` for rule_ids, so it would also be consistent.

File: includes/admin/class-cfwc-ajax.php [hunk 3]
```diff
@@ -117,9 +145,11 @@ public function delete_rule() {
 			wp_die( -1 );
 		}
 
-		$rule_id = isset( $_POST['rule_id'] ) ? absint( $_POST['rule_id'] ) : null;
+		// Support both string rule_id (new) and numeric index (legacy).
+		// phpcs:ignore WordPress.Security.ValidatedSanitizedInput.MissingUnslash, WordPress.Security.ValidatedSanitizedInput.InputNotSanitized
+		$raw_rule_id = isset( $_POST['rule_id'] ) ? wp_unslash( $_POST['rule_id'] ) : null;
```
Comment (dustinparker) [inline includes/admin/class-cfwc-ajax.php]:
```suggestion
		$raw_rule_id = isset( $_POST['rule_id'] ) ? sanitize_text_field( wp_unslash( $_POST['rule_id'] ) ) : '';
```

File: includes/admin/class-cfwc-ajax.php [hunk 4]
```diff
@@ -130,18 +160,37 @@ public function delete_rule() {
...
+
+		// Try matching by string rule_id first.
+		if ( is_string( $raw_rule_id ) ) {
+			foreach ( $rules as $index => $existing_rule ) {
+				if ( isset( $existing_rule['rule_id'] ) && $existing_rule['rule_id'] === $raw_rule_id ) {
+					$rule_index = $index;
+					break;
+				}
+			}
+		}
```
Comment (dustinparker) [inline includes/admin/class-cfwc-ajax.php]:
```suggestion
		foreach ( $rules as $index => $existing_rule ) {
			if ( isset( $existing_rule['rule_id'] ) && $existing_rule['rule_id'] === $raw_rule_id ) {
				$rule_index = $index;
				break;
			}
		}
```

Comment (dustinparker) [reply includes/admin/class-cfwc-ajax.php]: Since we sanitize as a text field above, we don't need the string check

**Review by dustinparker — CHANGES_REQUESTED** [review]
I've left a few suggestions/comments. Also, I don't see any interface for running rule import/exports, could you please explain how to test that functionality. Thanks!

**Comment by Abdalsalaam** [comment]
> I've left a few suggestions/comments. Also, I don't see any interface for running rule import/exports, could you please explain how to test that functionality. Thanks!

@dustinparker there's currently no UI for rules import/export yet, I only changed the import/export code in this PR to make sure the new changes were reflected there too not to ship the feature.

@shameemreza could you weigh in on the plan for the rules import/export UI? Is that something we want to build out as a follow-up, or should the backend handlers be removed until there's a UI to back them?

File: includes/admin/class-cfwc-ajax.php (hunk 1 above)
Comment (Abdalsalaam) [reply includes/admin/class-cfwc-ajax.php:401]: Good catch — fixed in a08efbe. The importer now mirrors the same `country` / `to_country` validation used in the settings save path: any CSV row with both fields empty is now skipped (`continue`) before being added to the import set, so empty cells can't produce inert rules anymore.

File: includes/class-cfwc-calculator.php (hunk 2 above)
Comment (Abdalsalaam) [reply includes/class-cfwc-calculator.php]: Agreed and fixed in c5f6c8e. Switched both wp_rand fallbacks in the calculator to `wp_generate_uuid4()` — the 4-digit random space was small enough that two unresolvable rules in the same cart calc could collide and silently overwrite each other in the `$calculated` map. UUIDs make that impossible and also keep the convention consistent with the rest of the codebase (settings save path and import already use it).

File: includes/admin/class-cfwc-ajax.php (hunk 3 above)
Comment (Abdalsalaam) [reply includes/admin/class-cfwc-ajax.php]: Applied the suggestion in 4168948. `$_POST['rule_id']` is now read with `sanitize_text_field( wp_unslash( ... ) )`, the empty-state default is `''` instead of `null`, and the `phpcs:ignore` for missing-unslash/not-sanitized is no longer needed.

File: includes/admin/class-cfwc-ajax.php (hunk 4 above)
Comment (Abdalsalaam) [reply includes/admin/class-cfwc-ajax.php]: Applied in 4168948 — since `sanitize_text_field( wp_unslash( ... ) )` already guarantees a string, the `is_string()` wrapper around the foreach is redundant and has been removed. The numeric fallback below still handles the legacy index case via `absint()`.

**Comment by CezaryDrewniak** [comment]
PR passes all provided test steps and handles additional edge cases (mixed-origin carts) correctly. The round-based dependency resolver and per-rule valuation logic are solid.

**Two maintenance concerns worth addressing before merge:**

**1. Orphaned dependencies after rule deletion**
If Rule B depends on Rule A via `base_includes` and Rule A is deleted, Rule B’s base_includes still retains the non-existent ID. The calculator silently ignores missing dependencies (it filters against matched rules and falls back to the rule’s own base), but the merchant receives no indication that a dependency is broken.

Suggestion: On rule deletion, scan remaining rules and remove the deleted rule_id from their base_includes. Alternatively, surface a validation notice in the rules table when a dependency target is missing.

**2. Dependency-cycle transient never cleared after fix**
set_transient`( 'cfwc_rules_dependency_error', ... )` is only deleted when` admin_notices()` runs. If a user fixes the cycle without loading an admin page, e.g., via CSV import, bulk settings save, or direct DB edit the stale transient persists indefinitely and could flash a false warning on the next admin visit.

Suggestion: Delete the transient inside` save_rules()`, `save_rule()`, `delete_rule()`, and `import_rules()` whenever the rule set is mutated, so the error flag is reset whenever the data changes.

File: includes/class-cfwc-calculator.php [hunk 5]
```diff
@@ -299,51 +297,148 @@ public function calculate_fees( $cart ) {
...
+			$rounds     = 0;
+			$max_rounds = count( $matching_rules ) + 1;
+
+			while ( ! empty( $pending ) && $rounds < $max_rounds ) {
+				++$rounds;
+				$made_progress = false;
+				$still_pending = array();
+
+				foreach ( $pending as $rule ) {
+					$rule_id = isset( $rule['rule_id'] ) && '' !== $rule['rule_id'] ? $rule['rule_id'] : (string) array_search( $rule, $matching_rules, true );
```
Comment (dustinparker) [inline includes/class-cfwc-calculator.php]: @Abdalsalaam if an item doesn't have a `rule_id` and the `array_search` does find a match, shouldn't we surface that to the merchant in the debug/error log so that they see that there are two identical rules?

Comment (Abdalsalaam) [reply includes/class-cfwc-calculator.php]: Good catch, fixed in b32a208 — calculator now logs a warning when two rules without a rule_id collide on the array_search fallback, so duplicates don't get silently overwritten in $calculated.

**Comment by Abdalsalaam** [comment]
Thanks @CezaryDrewniak, both fixed:

1. **Orphaned dependencies** — 0d65a05: `delete_rule` now scrubs the deleted `rule_id` from every other rule's `base_includes` before saving, so deps don't dangle.
2. **Stale cycle transient** — 24723c2: `cfwc_rules_dependency_error` is now cleared alongside the calculator cache in `save_rule`, `delete_rule`, `reorder_rules`, `import_rules`, and the settings save path.

File: includes/class-cfwc-settings.php [hunk 6]
```diff
@@ -271,13 +281,67 @@ private function save_rules() {
...
+
+		// Clear in-memory calculator cache for this request.
+		if ( class_exists( 'CFWC_Calculator' ) ) {
+			CFWC_Calculator::clear_cache();
+		}
+
+		// Reset the cycle-error flag — the rule set just changed, so any
+		// previous dependency-error notice is stale and should be re-derived
+		// on the next calculator run rather than flashed against new data.
+		delete_transient( 'cfwc_rules_dependency_error' );
```
Comment (dustinparker) [inline includes/class-cfwc-settings.php]: Clearing this transient unconditionally on save means a still-cyclic rule set will look "fixed" in admin until someone hits the cart again, since the warning only gets re-set when the calculator runs.

Could we re-run `CFWC_Calculator::has_cycle()` against the saved rules here and set or delete the transient based on the result, so the notice reflects the current state?

Comment (Abdalsalaam) [reply includes/class-cfwc-settings.php]: 🤔 let me check this

File: includes/class-cfwc-calculator.php [hunk 7]
```diff
@@ -299,51 +297,163 @@ public function calculate_fees( $cart ) {
...
+			if ( ! empty( $cycle_labels ) ) {
+				$cycle_labels = array_values( array_unique( $cycle_labels ) );
+				$this->log_debug(
+					sprintf(
+						'    WARNING: Cycle detected in base_includes for rules: %s. Skipping dependency fees for affected rules.',
+						implode( ', ', $cycle_labels )
+					),
+					'warning'
+				);
+				set_transient( 'cfwc_rules_dependency_error', $cycle_labels, DAY_IN_SECONDS );
```
Comment (dustinparker) [inline includes/class-cfwc-calculator.php]: Hey @Abdalsalaam, the `set_transient( 'cfwc_rules_dependency_error', ... )` call runs from `calculate_fees()`, which is hooked to `woocommerce_cart_calculate_fees`.

On a store with broken rules, that means every cart calculation performs a `wp_options` row write. This can happen on:

- Cart page load
- Checkout page load
- Mini-cart refresh
- Shipping method change
- Coupon application

A single shopper going through checkout could trigger this 5-10 times. Can we remove this or consider other approaches?

File: includes/class-cfwc-calculator.php [hunk 8]
```diff
@@ -678,6 +788,137 @@ private function get_all_rules() {
...
+			}
+			$path   = $visited;
+			$path[] = $dep_id;
+			if ( self::has_cycle( $rules, $dep_id, $path ) ) {
+				return true;
+			}
+		}
+
+		return false;
+	}
```
Comment (dustinparker) [inline includes/class-cfwc-calculator.php:930]:
```suggestion
	/**
	 * Detect whether a rule participates in a `base_includes` cycle.
	 *
	 * Walks the dependency graph breadth-first from `$start_id` and returns
	 * true only if a path leads back to `$start_id` itself — i.e. `$start_id`
	 * is part of a cycle. Rules that are merely upstream of a cycle (they
	 * reach a cycle but do not loop back to themselves) correctly return
	 * false; the calculator's round-loop already handles their unresolvable
	 * dependencies via the own-base fallback.
	 *
	 * @since 1.2.0
	 * @param array  $rules    Rules to check.
	 * @param string $start_id Rule ID to test for cycle membership.
	 * @return bool True if `$start_id` participates in a cycle.
	 */
	public static function has_cycle( $rules, $start_id ) {
		$rule_map = array();
		foreach ( $rules as $rule ) {
			$rid = isset( $rule['rule_id'] ) ? $rule['rule_id'] : '';
			if ( ! empty( $rid ) ) {
				$rule_map[ $rid ] = $rule;
			}
		}

		if ( ! isset( $rule_map[ $start_id ] ) ) {
			return false;
		}

		$rule  = $rule_map[ $start_id ];
		$queue = isset( $rule['base_includes'] ) && is_array( $rule['base_includes'] )
			? $rule['base_includes']
			: array();
		$seen  = array();

		while ( ! empty( $queue ) ) {
			$current = array_shift( $queue );

			if ( $current === $start_id ) {
				return true;
			}
			if ( isset( $seen[ $current ] ) ) {
				continue;
			}
			$seen[ $current ] = true;

			if ( ! isset( $rule_map[ $current ] ) ) {
				continue;
			}

			$cur_rule = $rule_map[ $current ];
			$cur_deps = isset( $cur_rule['base_includes'] ) && is_array( $cur_rule['base_includes'] )
				? $cur_rule['base_includes']
				: array();
			foreach ( $cur_deps as $dep ) {
				$queue[] = $dep;
			}
		}

		return false;
	}
```

Comment (dustinparker) [reply includes/class-cfwc-calculator.php:930]: @Abdalsalaam what do you think of changing this to only show the merchant the rules that have cyclical dependencies directly rather than also showing them higher up rules that don't actually have cyclical dependencies themselves?

File: includes/class-cfwc-templates.php
Comment (dustinparker) [inline includes/class-cfwc-templates.php:1]: Hey @abdalsalaam,

If I append a compound preset over an existing general rule, it looks like the new compound rule gets dropped, but the UI still reports success. I noticed this in `apply_template()` append mode in `class-cfwc-templates.php` around lines 1421-1429.

The current "conflicting general rules" check treats any incoming general rule as a duplicate if a general rule already exists for that destination, regardless of the rule label. That works for simple presets, but it breaks compound presets. For example, the Canada GST & Duty preset needs two general rules for the same country: one rule for Canada Duty (Import), and another rule for Canada GST.

If the merchant already has the duty rule and then appends the Canada GST & Duty preset, the GST rule gets treated as a duplicate of the existing duty rule and is skipped. The user does not see an error. They get this success message instead:

> All preset rules already exist. No rules added.

That is misleading because the GST rule does not already exist. It was discarded. So an existing merchant trying to adopt the updated preset is told they are already set up correctly, but they remain missing the GST rule.

To reproduce this, add one general Canada rule, such as Canada Duty (Import) at 8% for all products. Then append the Canada GST & Duty preset. The **expected result** is that the Canada GST rule should be added and set to depend on the existing duty rule. The **actual result** is that no rule is added, and the user sees the message above.

The recommended fix would be to only skip an incoming general rule when both the country and the label match an existing rule. Also, when the duty rule is a real duplicate, we should remap the GST rule's `base_includes` value to the existing duty rule's `rule_id`. The `$id_remap` rewrite at `class-cfwc-templates.php:1393-1400` already handles this for collision cases, but it does not currently cover the skipped-duplicate case.

WDYT?

**Review by dustinparker — CHANGES_REQUESTED** [review]

File: includes/class-cfwc-templates.php
Comment (Abdalsalaam) [reply includes/class-cfwc-templates.php:1]: Thanks for catching this!
I was able to replicate the issue and I'm working on it

**Review by CezaryDrewniak — APPROVED** [review]
LGTM!

---

## Test Results Summary

✅ **Core functionality works** – all preset calculations match expected values.  

Tests done with $100 product and $20 shipping.

| Test Case | Expected Result | Actual Result | Status |
|-----------|----------------|---------------|--------|
| Canada preset (duty FOB, GST CIF + duty) | Duty $8, GST $6.40 | Duty $8, GST $6.40 | ✅ Pass |
| Australia preset (duty FOB, GST CIF + duty) | Duty $5, GST $12.50 | Duty $5, GST $12.50 | ✅ Pass |
| New Zealand preset (duty FOB, GST CIF + duty) | Duty $5, GST $18.75 | Duty $5, GST $18.75 | ✅ Pass |
| UK / EU preset (both CIF, VAT compounds on duty) | Duty $6, VAT $25.20 | Duty $6, VAT $25.20 | ✅ Pass |
| Self‑dependency prevention | Rule not listed in its own “Depends on” dropdown | Not selectable | ✅ Pass |
| Cycle detection (A → B → A) | Admin warning appears | Warning appears only after **Save changes** (not on inline edit) and is low specifity (Customs Fees Warning: rule-name, rule-name) | ✅ Pass, but warning could be higher specifity |
| Per‑rule valuation column | Shows “Global” or override badge | Visible and correct | ✅ Pass |
| “Depends on” column | Shows “+1 fee” with tooltip | Works as expected | ✅ Pass |
| Migration from v1.1.x | `rule_id`, `valuation_method`, `base_includes` added | Migrated successfully | ✅ Pass |

**Review by dustinparker — APPROVED** [review]
The rules with overrides/stacking/priority are working as expected. Looks good!
